plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/weighers/DeprecatedWeigher.kt

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+package org.jetbrains.kotlin.idea.completion.weighers
+
+import com.intellij.codeInsight.lookup.LookupElement
+import com.intellij.codeInsight.lookup.LookupElementWeigher
+import com.intellij.openapi.util.Key
+import org.jetbrains.kotlin.idea.frontend.api.KtAnalysisSession
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtSymbol
+import org.jetbrains.kotlin.psi.UserDataProperty
+
+internal object DeprecatedWeigher {
+    const val WEIGHER_ID = "kotlin.deprecated"
+    private var LookupElement.deprecated by UserDataProperty(Key<Boolean>("KOTLIN_DEPRECATED"))
+
+    fun KtAnalysisSession.addWeight(lookupElement: LookupElement, symbol: KtSymbol) {
+        // Here we use an if branch so that we don't store a pointer to `false` for elements that are not deprecated.
```

Comment:
You can avoid that `if` at all if you use `NotNullableUserDataProperty` with default `false` value - it is optimized exactly for such case

Reply:
Nice! I didn't know this exists!